packages/git/src/browser/history/git-history-widget.ts

```diff
@@ -0,0 +1,417 @@
+import { GitFileChangeNode } from "../git-widget";
+import { SelectionService } from "@theia/core";
+import { Message } from "@phosphor/messaging";
+import { ElementExt } from "@phosphor/domutils";
+import { FileSystem } from "@theia/filesystem/lib/common";
+import { Key } from "@theia/core/lib/browser/keys";
+
+export interface GitCommitNode {
+    readonly authorName: string;
+    readonly authorEmail: string;
+    readonly authorDate: Date;
+    readonly authorDateRelative: string;
+    readonly commitMessage: string;
+    readonly messageBody?: string;
+    readonly fileChangeNodes: GitFileChangeNode[];
+    readonly commitSha: string;
+    expanded: boolean;
+    selected: boolean;
+}
+
+export namespace GitCommitNode {
+    export function is(node: any): node is GitCommitNode {
+        return 'commitSha' in node && 'commitMessage' in node && 'fileChangeNodes' in node;
+    }
+}
+
+export enum SelectDirection {
+    NEXT, PREVIOUS
+}
+
+export type GitHistoryListNode = (GitCommitNode | GitFileChangeNode);
+
+@injectable()
+export class GitHistoryWidget extends GitBaseWidget implements StatefulWidget {
+    protected options: Git.Options.Log;
+    protected commits: GitCommitNode[];
+    protected historyList: GitHistoryListNode[];
+    protected ready: boolean;
+    protected singleFileMode: boolean;
+
+    constructor(
+        @inject(GitRepositoryProvider) protected readonly repositoryProvider: GitRepositoryProvider,
+        @inject(LabelProvider) protected readonly labelProvider: LabelProvider,
+        @inject(OpenerService) protected readonly openerService: OpenerService,
+        @inject(SelectionService) protected readonly selectionService: SelectionService,
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(Git) protected readonly git: Git) {
+        super(repositoryProvider, labelProvider);
+        this.id = GIT_HISTORY;
+        this.title.label = "Git History";
+        this.addClass('theia-git');
+
+        this.node.tabIndex = 0;
+
+        selectionService.onSelectionChanged((c: GitHistoryListNode) => {
+            c.selected = true;
+            this.update();
+        });
+    }
+
+    protected onUpdateRequest(msg: Message): void {
+        super.onUpdateRequest(msg);
+
+        const selected = this.node.getElementsByClassName(SELECTED_CLASS)[0];
+        const scrollArea = document.getElementById('git-history-list-container');
+        if (selected && scrollArea) {
+            ElementExt.scrollIntoViewIfNeeded(scrollArea, selected);
+        }
+    }
+
+    async setContent(options?: Git.Options.Log) {
+        this.options = options || {};
+        this.commits = [];
+        this.ready = false;
+        if (options && options.uri) {
+            const fileStat = await this.fileSystem.getFileStat(options.uri);
+            this.singleFileMode = !fileStat.isDirectory;
+        }
+        this.addCommits(options);
+        this.update();
```

Comment:
this update looks like it should update after `addCommits` but it doesn't because `addCommits` is async.

Reply:
but I see it is there to render the loading